Allow lua to be used in runtime files

shadmansaleh opened the pull request:
### This pr adds lua support to runtime files

Currently added:

- [x] colors
- [x] compiler
- [x] ftplugin
- [x] ftdetect
- [x] indent
- [x] plugin
- [x] syntax


### Behavior

I've used whether multiple same named vim files executed or not as standard for whether to run lua files along with viml files or run first one found . For example if  `c.vim` files exists in two location in runtime and both gets executed then all `c.lua` files will be executed too . If the first `c.vim` in runtime path gets executed and rest get ignored then `c.lua` will be ignored too [ This is is becuase `.vim` files always takes precedence over `.lua` files . So in any case `.vim` files will be executed before `.lua` files.

- colors     [First found]
- compiler [First found]
- ftplugin   [All]
- ftdetect   [All | Ran at startup or `packadd`] 
- indent      [All]
- plugin      [All | Ran at startup or `packadd`]
- syntax     [All]


### Aditional changes due to this pr
Since `do_source()` function now recognises lua files . `:runtime` & `:source` command executes lua files currectly now :D 

#### Comment:
With lua plugin echosystem growing everyday I think this will be a nice addition . 

closes #12670

shadmansaleh committed:
feat(startup): Source runtime/plugin/**/*.lua at startup

For opt plugins these files are sourced on `:packadd`

* `:runtime` Now can exexute lua files

clason commented:
I think this is something that needs to be discussed in detail first; starting to execute globs of Lua can be dangerous!

Related issue (and linked discussion): https://github.com/neovim/neovim/issues/12670

shadmansaleh commented:
> I think this is something that needs to be discussed in detail first

Definatly . I thought opening a pr might make it faster then an issue :)

> starting to execute globs of Lua can be dangerous

Can you elaborate ? I'm curious to know what danger it adds since plugin/**/*.vim files are currently executed in this manner. Is this something lua specific ?

> I'd definitely omit packpath and start with stdpath('config')/plugin etc.

If packpath is ommited plugins won't be able to use it . It kind of makes this useless IMO.

> first -- ideally in a new directory with a clear name (lua_start?)

That can definitely be done . I'm not sure necessary. If we keep adding lua specific folders won't it make runtime messier ? Since extensions are different file_names can't collide . Should I make the change or wait for a 3rd opinion ? 

> Related issue (and linked discussion): #12670

Thanks for sharing the issue . I don't know how I missed it . I'll read through older discussions :)

clason commented:
Maybe I just don't understand fully what exactly is executed after this PR. I'm a bit worried that this can execute Lua files in plugins (installed as packages) that are not meant to be executed directly.

Only acting on `stdpath('config')` is definitely not useless since it'll allow `ftplugin`s and allow putting Lua config in `plugin` and not need a vimscript shim just to `require` them.

clason commented:
But I'll let @tjdevries do most of the talking here since he's already got an implementation in https://github.com/tjdevries/astronauta.nvim

shadmansaleh commented:
> Maybe I just don't understand fully what exactly is executed after this PR. I'm a bit worried that this can execute Lua files in plugins

No it can't . I think the word plugin is confusing here . AFAIK a plugin folder say named `astronauta.nvim` is added to rtp . This plugin will execute lua files in `astronauta.nvim/plugin` folder . Currently `.vim` files in that folder get executed.

Note: Currently plugins store lua files under `astronauta.nvim/lua/` those won't be touched . And ftplugins aren't handled by this pr . I was thinking of doing a follow up pr based on what we discuss here.

> https://github.com/tjdevries/astronauta.nvim

I've looked at the source of that plugin . It basicaly does the same thing glob sources all `plugin/**/*.lua` file. See [here](https://github.com/tjdevries/astronauta.nvim/blob/e69d7bdc4183047c4700427922c4a3cc1e3258c6/lua/astronauta/plugin.lua#L10)

clason commented:
> No it can't . I think the word plugin is confusing here . Say a plugin folder named `astronauta.nvim` is in rtp . This plugin will execute lua files in `astronauta.nvim/plugin` folder . Currently `.vim` files in that folder get executed.
> 
> Note: Currently plugins store lua files under `astronauta.nvim/lua/` those won't be touched . And ftplugins aren't handled by this pr . I was thinking of doing a follow up pr based on what we discuss here.

That was exactly my worry, so consider me reassured!
 
And I think adding the other "special" directories (`after`, `ftplugin`, `colors`, ...) one by one is worthwhile.

shadmansaleh commented:
> first -- ideally in a new directory with a clear name (lua_start?)

@clason seeing older discussions [comment1](https://github.com/neovim/neovim/issues/7895#issuecomment-633152541) [comment2](https://github.com/neovim/neovim/issues/7895#issuecomment-633157289) [comment3](https://github.com/neovim/neovim/issues/7895#issuecomment-633985681)

It seems using same `plugin/` folder and sources `*.vim` first then `*.lua`  was agreed upon.

clason commented:
Yep, sorry, forgot to comment on that explicitly: vim first, then lua sounds like the best option to me. (Definitely should be all of one then all of other, not alphabetically... :])

clason commented:
But the special case that both `foo.vim` and `foo.lua` exists should be discussed (and documented) explicitly. I don't expect this to be a problem in general but one can probably come up with hypothetical examples where they could interfere (plugin trying to support both vim/nvim 0.4 and nvim 0.5+?)

shadmansaleh commented:
> But the special case that both foo.vim and foo.lua exists should be discussed (and documented) explicitly.

I think they should be treated same as different filenames . Explicitly handling same names will add confusion and complication IMO. 

>I don't expect this to be a problem in general but one can probably come up with hypothetical examples where they could interfere (plugin trying to support both vim/nvim 0.4 and nvim 0.5+?)

If author of plugin wants to support older version I think he/she'll have to write shims or use older way (For example load .lua from .vim like currently/ Put version guards) . I think this is something plugin authors already do . So doesn't feel like it'll be a problem .

About documentation : I think *.vim files being loaded at startup was mentiond somewhere on vimdoc . But I don't remember where it was and greping isn't being much of a help (I get too many hits) . Would definatly love if someone could tell me where it was if it was at all :)

clason commented:
> I think they should be treated same as different filenames . Explicitly handling same names will add confusion and complication IMO.

Agreed. If both `foo.lua` and `foo.vim` exist, both should be sourced: `foo.vim` first, then (after the rest of the `vim` files)  `foo.lua`. I just wanted to bring it up so it's in the open (and documented).

> If author of plugin wants to support older version I think he/she'll have to write shims or use older way (For example load .lua from .vim like currently/ Put version guards) . I think this is something plugin authors already do . So doesn't feel like it'll be a problem .

As I said, a constructed example.

> About documentation : I think *.vim files being loaded at startup was mentiond somewhere on vimdoc . But I don't remember where it was and greping isn't being much of a help (I get too many hits) . Would definatly love if someone could tell me where it was if it was at all :)

`:h startup` would be the first idea.

bfredl commented on src/nvim/runtime.c:
```diff
@@ -405,31 +415,39 @@ static int add_pack_dir_to_rtp(char_u *fname)
   size_t len = strlen(ffname) + STRLEN(ftpat);
-  char_u *pat = try_malloc(len + 1);
-  if (pat == NULL) {
+  char_u *pat_vim = try_malloc(len + 1);
```
can use `xmallocz(len)` and skip the NULL checks.

shadmansaleh replied:
`xmallocz` Looks right . One thing though currently `NULL` is recived ffname is freed (threadend at line 453) . We can get rid of that too . NULL means we are out of memory and are crashing . Those leaked memory will be reclaimed by system when the prossess exits . So freeing isn't nessary right ?

shadmansaleh replied:
Also currently `FAIL` is returned only when those memory allocation fails . If those are that's handled by `xmalloc` then return value of this4 function has no significance as it'll always return `OK` and when memory allocation fails nvim get closed . 

Is current setup some how allow to survive out of memory error ?

shadmansaleh replied:
@bfredl updated. Now using `xmallocz`

shadmansaleh commented:
> :h startup would be the first idea.

@clason seems appropriate . I've added docs.

I can't think of anything else to do on thi pr . Now wait for reviewers 😴

tjdevries commented:
I'm not sure if it's worthwhile adding to the docs, but I think it might be nice to let people know that these files are not executed like they woudl be with `require` (i.e. re-executing them actually runs the code again).

One question: does `source ~/path/to/file.lua` execute the file as a lua file now? Another question: Does `runtime! plugin/my_thing.lua` also work?

If so, we should update the documentation for both `:source` and `:runtime`

shadmansaleh commented:
> I'm not sure if it's worthwhile adding to the docs

Why not ? I actually like the fact about vimdocs that if you're looking for something you have a good chance of finding the details there.

> but I think it might be nice to let people know that these files are not executed like they woudl be with require (i.e. re-executing them actually runs the code again).

Definatly . I'll add that it's like luafile not like require.

> One question: does source ~/path/to/file.lua execute the file as a lua file now? Another question: Does runtime! plugin/my_thing.lua also work?

`source` no, `runtime` yes. source follows a different code path . I'll update the docs for runtime.

 Is `source` running lua files a wanted behavior ?  That makes `luafile` obsolete . TBH I'm now wondering why this wasn't added in the first place instead of adding new `luafile` command.

shadmansaleh commented:
@tjdevries I've updated the docs for `:runtime` and have added how lua files are spurced there . Since other doc updates say they use `:runtime` I think it's fine to add it there .

tjdevries commented:
> source no, runtime yes. source follows a different code path . I'll update the docs for runtime.

> Is source running lua files a wanted behavior ? That makes luafile obsolete . TBH I'm now wondering why this wasn't added in the first place instead of adding new luafile command.

@bfredl do you remember anything about this?

clason commented:
Just a short heads-up: It makes more sense to do this in concert with the remaining config directories and not shoe-horn it into the 0.5 release. (This has nothing to do with this specific PR, which is in good shape already 👍🏻 ) This will also give it a bit more visibility in the release notes ;)

bfredl commented:
@tjdevries luafile was just taken from if_lua along with the other entry points. This was years before any discussion of lua replacing any runtime plugin.

shadmansaleh commented:
@clason 0.5.1 is pretty far away I might Add other runtime items to recognize lua files in this pr by then .
Does all the possible runtime directories exist is /runtime is source tree ?
I'm thinking these might execept lua files
- colors
- compiler
- ftplugin
- indent
- keymap
- plugin
- syntax

Is there something to add or remove from this list ?

shadmansaleh committed:
feat(runtime): Allow lua to be used in colorschemes

* tests(runtime): move runtime/plugin tests to functional/lua/runtime_spec

clason commented:
> @clason 0.5.1 is pretty far away 

No, not necessarily; I deliberately milestoned it for the "very next (bugfix) release after 0.5", not the "next major release (aka when tree-sitter is stable)".

The point is that we don't want to hold up 0.5 for this, and it would be good to have it bake a bit on master before pushing it out to a release to catch possible unwanted interactions (especially with plugins in the wild, so they have a bit of time to adapt).

But I am not averse to merging it if it's ready a good while before the remaining (few) blockers for 0.5 are fixed.

> I might Add other runtime items to recognize lua files in this pr by then .

That would be good in any case; saves us from a weird intermediate state where some things work and some don't :)

> Does all the possible runtime directories exist is /runtime is source tree ?
> I'm thinking these might execept lua files
> 
> * colors
> * compiler
> * ftplugin
> * indent
> * keymap
> * plugin
> * syntax
> 
> Is there something to add or remove from this list ?

`ftdetect` also comes to mind. @bfredl ?

shadmansaleh commented:
>  ftdetect also comes to mind.

@clason Yes .  that can be included . Do you agree keymap and syntax not being lua ?

clason commented:
Hmm, I personally don't use them, so can't say. My gut feeling is that it's simpler to just allow Lua everywhere you allow vim, even if it doesn't really make much sense to use Lua over vimscript -- because when has that ever stopped someone from trying to do it or asking for it?

Also makes it easier to document: no list of exceptions, just "Lua everywhere".

shadmansaleh commented:
Can some one point me to the direction how completion for commands are done ?

Now `:colorscheme` command can load lua colorscheme but it doesn't provide completion for them .

@bfredl any idea where I should look ?

shadmansaleh committed:
feat(runtime): Allow lua to be used in compiler

shadmansaleh committed:
feat(runtime): Allow lua to be used in ftplugin

shadmansaleh committed:
feat(runtime): Allow lua to be used in indent

shadmansaleh committed:
feat(runtime): Allow lua to be used in ftdetect

clason commented:
From `:h 'rtp'`:
```
 	List of directories to be searched for these runtime files:
	  filetype.vim	filetypes by file name |new-filetype|
	  scripts.vim	filetypes by file contents |new-filetype-scripts|
	  autoload/	automatically loaded scripts |autoload-functions|
	  colors/	color scheme files |:colorscheme|
	  compiler/	compiler files |:compiler|
	  doc/		documentation |write-local-help|
	  ftplugin/	filetype plugins |write-filetype-plugin|
	  indent/	indent scripts |indent-expression|
	  keymap/	key mapping files |mbyte-keymap|
	  lang/		menu translations |:menutrans|
	  menu.vim	GUI menus |menu.vim|
	  pack/		packages |:packadd|
	  plugin/	plugin scripts |write-plugin|
	  print/	files for printing |postscript-print-encoding|
	  rplugin/	|remote-plugin| scripts
	  spell/	spell checking files |spell|
	  syntax/	syntax files |mysyntaxfile|
	  tutor/	tutorial files |:Tutor|
```

(Not all of them make sense or are relevant, but if in doubt, might as well source Lua from there, even if there should be none -- for now.)

Plus `after`, which is important to support as well.

shadmansaleh committed:
enhance(runtime): Enable completion for lua files

Enabled for `:colorscheme` `:compiler` filetype

shadmansaleh commented:
@clason I've seen that list . And currently this pr has support for everything except `syntax` . I can't find the entry point for that lol .

> Plus after, which is important to support as well.

after has same directory structure as any normal rpt dir (like /runtime) Only speciality being things in there gets executed after normal rtp so it's already supported.

Also CI seems stuck can you kill older jobs from this pr ? I wish github would just allow me to kill my jobs without repo access :D .

shadmansaleh committed:
feat(runtime): Allow lua to be used in syntax

tjdevries commented on src/nvim/digraph.c:
```diff
@@ -1833,27 +1833,33 @@ char_u* keymap_init(void)
+      "keymap/%s_%s.lua",
+      "keymap/%s.vim",
+      "keymap/%s.lua",
+    };
```
So I'm not sure that keymap makes sense for lua (haven't read all the comments yet) but it doesn't really execute `vimscript` in keymap files, right?

shadmansaleh replied:
Well I don't think keymap makes sense either . Added because @clason said to :P

clason replied:
That's right, blame me :P 

But you've actually already convinced me that `keymap` doesn't make sense here (no vim => no Lua)

shadmansaleh replied:
So we agree on removing keymap ?

clason replied:
Yes. Also `autoload` makes little sense since the `foo#bar` syntax isn't easy to support -- we can punt to `/lua` and `require('foo').bar()` for that.

shadmansaleh replied:
Ok. I've reverted the keymap commit.

Also @clason `autoload` is in this pr . It's very vim specific . So I agree it can't be lua :)

tjdevries commented on src/nvim/ex_cmds2.c:
```diff
@@ -2420,7 +2420,8 @@ void ex_compiler(exarg_T *eap)
     // List all compiler scripts.
-    do_cmdline_cmd("echo globpath(&rtp, 'compiler/*.vim')");  // NOLINT
+    // [vl][iu][ma] = lua | vim
+    do_cmdline_cmd("echo globpath(&rtp, 'compiler/*.[vl][iu][ma]')");  // NOLINT
```
won't this also match `lim` or other random combinations?

shadmansaleh replied:
It actually will. I'll change that

shadmansaleh replied:
resolved

tjdevries commented on src/nvim/main.c:
```diff
@@ -1367,7 +1367,8 @@ static void load_plugins(void)
   if (p_lpl) {
     char_u *rtp_copy = NULL;
-    char_u *const plugin_pattern = (char_u *)"plugin/**/*.vim";  // NOLINT
+    char_u *const plugin_pattern_vim = (char_u *)"plugin/**/*.vim";  // NOLINT
```
Is there no way to do `*.(vim|lua)` or other -- I don't remember which matching is used here :laughing:

shadmansaleh replied:
No . It doesn't work I've tried that

clason commented:
> @clason I've seen that list . 

Just wanted to make sure :)

> And currently this pr has support for everything except `syntax` . I can't find the entry point for that lol .

@bfredl will know, I'm sure :]

> after has same directory structure as any normal rpt dir (like /runtime) Only speciality being things in there gets executed after normal rtp so it's already supported.

Great!

> #### Also CI seems stuck can you kill older jobs from this pr ?

I cannot, but @bfredl might.

> I wish github would just allow me to kill my jobs without repo access :D .
> 
> Also is it possible to have a timeout for CI jobs so they get killed automatically if they take too long to complete

Not that I know of, at least not individually.

tjdevries commented on src/nvim/runtime.c:
```diff
@@ -49,7 +50,11 @@ void ex_runtime(exarg_T *eap)
 
 static void source_callback(char_u *fname, void *cookie)
 {
-  (void)do_source(fname, false, DOSO_NONE);
+  if (path_with_extension((const char *)fname, "lua")) {
+    nlua_exec_file((const char *)fname);
+  } else {
+    (void)do_source(fname, false, DOSO_NONE);
+  }
```
what if you change `do_source` to be the one that handles checking the path?

I see this exact if statement more than once in the code, which is worrisome that we'll miss something.

shadmansaleh replied:
Those statements are used in 2 places . I didn't want to change `do_source` beacause it can affect a lot of other stuff that I'm probably not aware of. For example `:source` of lua files will actually work .

bfredl replied:
hmm my gut feeling is that _if_ `runtime some/file.lua` works then so should _also_ `source also/some/file.lua`

shadmansaleh replied:
Ok . I'll move the funtionality to `do_source` Since you both want it. Well it's an extension check . Recognizing lua files as lua files instead of vim can't break anything I think :)

shadmansaleh replied:
Ok done . Now you can do source on lua files too . I've looked at all the uses of do_source . It seems to be a harmless addition :)

jamessan replied:
> Ok . I'll move the funtionality to `do_source` Since you both want it. Well it's an extension check . Recognizing lua files as lua files instead of vim can't break anything I think :)

What if there's no extension on the file?

shadmansaleh replied:
Assumed vil like now

tjdevries reviewed (commented):
OK, so I was digging and it doesn't look like you can do `(vim|lua)` in the matches, so probably ignore that comment. I'll look some more later.

tjdevries commented:
hmmm... for some reason the tests are hanging for me locally....

Just hangs here:

```
[----------] Global test environment setup.
[----------] Running tests from test/functional/lua/runtime_spec.lua
[ RUN      ] runtime: plugin loads plugin/*.lua from XDG config home: 271.12 ms OK
```

shadmansaleh commented:
> hmmm... for some reason the tests are hanging for me locally....

Strange it works for me . CI seems to pass🤔

By the way I've added a new test files does anything need to done extra apart from adding the file for CI to run it?

tjdevries commented:
It will run automatically in CI (it searches for *_spec.lua files)

tjdevries commented:
I cannot figure out why this would be hanging on the tests... very weird. I will try on fresh install if nothing else works.

tjdevries commented:
What happens when you run this:

`TEST_FILE=test/functional/lua/runtime_spec.lua make functionaltest`

shadmansaleh commented:
I've been running it all day

Output
```log
❯ TEST_FILE=test/functional/lua/runtime_spec.lua make functionaltest
mkdir -p build
touch build/.ran-third-party-cmake
ninja -j8 -C .deps
ninja: Entering directory `.deps'
[3/3] cd /home/shadman/projects/neovim/neo...ps && /usr/bin/cmake -E touch .third-party
ninja -j8 -C build
ninja: Entering directory `build'
[6/6] Generating doc/tags
ninja -j8 -C build functionaltest
ninja: Entering directory `build'
[1/2] cd /home/shadman/projects/neovim/neo...rojects/neovim/neovim/cmake/RunTests.cmake
[----------] Global test environment setup.
[----------] Running tests from test/functional/lua/runtime_spec.lua
[ RUN      ] runtime: plugin loads plugin/*.lua from XDG config home: 438.17 ms OK
[ RUN      ] runtime: plugin loads plugin/*.lua from start plugins: 323.33 ms OK
[ RUN      ] runtime: colors loads lua colorscheme: 166.73 ms OK
[ RUN      ] runtime: colors loads vim colorscheme when both lua and vim version exist:
338.74 ms OK
[ RUN      ] runtime: compiler loads lua compilers: 84.66 ms OK
[ RUN      ] runtime: compiler loads vim compilers when both lua and vim version exist:
358.00 ms OK
[ RUN      ] runtime: ftplugin loads lua ftplugins: 368.13 ms OK
[ RUN      ] runtime: indent loads lua indents: 358.03 ms OK
[ RUN      ] runtime: ftdetect loads lua ftdetects: 310.70 ms OK
[ RUN      ] runtime: keymap loads lua keymaps: 338.40 ms OK
[ RUN      ] runtime: syntax loads lua syntaxes on filetype change: 363.48 ms OK
[ RUN      ] runtime: syntax loads lua syntaxes on syntax change: 324.67 ms OK
[----------] 12 tests from test/functional/lua/runtime_spec.lua (4286.14 ms total)      
[----------] Global test environment teardown.
------------------------------------------------------------------------------          
------------------------------------------------------------------------------
[==========] 12 tests from 1 test file ran. (4287.77 ms total)
[  PASSED  ] 12 tests.
```

tjdevries commented on test/functional/lua/runtime_spec.lua:
```diff
@@ -0,0 +1,224 @@
+      mkdir_p(plugin_folder_path)
+      write_file(plugin_file_path, [[ vim.g.lua_plugin = 1 ]])
+
+      clear{ args_rm={'-u' }, env={ XDG_CONFIG_HOME=xconfig }}
```
I don't like using `clear` here to set a new nvim instance in the middle of a test -- it's a bit weird that `mkdir_p` communicates with the firts nvim, then you clear, and then now it's a new nvim as the session.

shadmansaleh replied:
double clear is kind of weird true . But the test that check for stuff happening at startup ( plugin & ftdetect here) need the second clear . The rest don't . I've removed it from them .

tjdevries commented:
very weird -- anyone else having the same problem?

shadmansaleh committed:
refactor(tests): Simplify tests at functional/lua/runtime_spec

shadmansaleh commented:
> very weird -- anyone else having the same problem?

I'd be very happy if you could pin point why / what actually getting stuck . I think some CI runs might be facing it too . A few of them are randomly stuck . I can't reproduce it on my local machine so I'm not sure what's causing them .

tjdevries commented:
I think the tests should be added to `./test/functional/core/startup_spec.lua` -- since that's the startup scenario.

The other ones can be done via `set rtp+=...` or `vim.opt.rtp:append { './path' }` and then using `runtime` commands or similar.

shadmansaleh commented:
So . Move `plugin` & `ftdetect` tests to `startup_spec` and leave the rest where they are ?

I like the idea about `rtp+`

tjdevries commented:
yes, `startup` logic should go in that `startup_spec` test file.

Things for doing stuff at runtime, like `runtime plugin/my_thing.lua` or whatever, can happen elsewhere. And you don't need to restart nvim in the middle of the tests with `clear`, but instead can just modify the runtime directly and ask nvim questions before & after your commands.

tjdevries commented:
( `clear` creates a NEW neovim instance and sets it as the current session, which is why I think it feels weird to do in the middle of a functional test, if that makes sense )

shadmansaleh commented:
I've updated the tests . 

Moved plugin & ftdetect tests to startup_spec,
and now using rtp instead of clear to set XDG_CONFIG_DIR.

Tests are dafinatly a lot faster without those clears .
@tjdevries can you check if the tests still gets stuck on your device ?

shadmansaleh committed:
refactor(source): Move lua file detection to do_source

So now :source can run lua files too :)

* feat: Add support for :[ranged]source for lua files

shadmansaleh commented:
@clason @bfredl how far is 0.5 actually ? If there are some time . This can probably be included in 0.5 . I think this pr is done unless someone reviewing raises an issue :)

tjdevries commented:
Hmmm... it looks like some stuff is still hanging for some reason -- but I think it's something to do with the way you're doing the tests (I just can't figure out why at this point yet). I will take a look at it some more hopefully on Friday

shadmansaleh commented:
> Hmmm... it looks like some stuff is still hanging for some reason

Which ones are hanging ?

I've checked the CI logs that were hanging . They were randomly hanging on unrealated tests (shell, lsp, ui...) randomly . It didn't make any sense to me .

shadmansaleh commented:
@bfredl, @clason or any one else can you try running the tests from this pr and see if they get stuck on your machine or not ?

```sh
TEST_FILTER=runtime TEST_FILE=test/functional/lua/runtime_spec.lua make functionaltest
TEST_FILTER=runtime TEST_FILE=test/functional/core/startup_spec.lua make functionaltest
```

clason commented:
@shadmansaleh sorry for the delay. I just ran these two test sets (8+3) on my machine (macOS Big Sur, Intel) and both passed.

shadmansaleh commented:
@clason thanks for running the tests. I really can't understand why they get stuck for @tjdevries 🤷‍♂️

shadmansaleh commented:
@kutsan  thanks. @tjdevries is it possible somthing on your end is causing the tests to get stuck ? Can you give it another try and let me know which tests are getting stuck and which aren't among those 11 ? It'll help us isolate the issue and find the source :)

clason commented:
@shadmansaleh btw, if you rebase on master, the macOS CI should pass now

shadmansaleh commented:
> btw, if you rebase on master, the macOS CI should pass now

Yes I know mac CI was fixed . I thought I'd rerun CI after sourcehut's openbsd CI got patched.

bfredl commented on runtime/doc/starting.txt:
```diff
@@ -469,10 +469,15 @@ accordingly.  Vim proceeds in this order:
+	then all the "*.lua" files will be sourced. If two files with same
+	name but different extensions exists they will be treated in same
+	manner. For example when both "foo.vim" and "foo.lua" exists then
+	first "foo.vim" will be sourced then "foo.lua" will be ran.
```
ran -> sourced

shadmansaleh replied:
done

bfredl commented on runtime/doc/repeat.txt:
```diff
@@ -172,8 +172,7 @@ Using Vim scripts					*using-scripts*
 					*:so* *:source* *load-vim-script*
-:so[urce] {file}	Read Ex commands from {file}.  These are commands that
-			start with a ":".
+:so[urce] {file}	Runs vim or lua {file}
```
perhaps mention if there is no extension, the file is assumed to be vim script?

shadmansaleh replied:
done

shadmansaleh commented on test/functional/core/startup_spec.lua:
```diff
@@ -494,6 +495,60 @@ describe('user config init', function()
+    write_file(ftdetect_file , [[vim.g.lua_ftdetect = 1]])
+
+    clear{ args_rm={'-u'}, env={XDG_CONFIG_HOME=xconfig,
+                                VIMRUNTIME='runtime/'}}
```
@bfredl any idea why this test fail if I don't set `VIMRUNTIME` env ?

bfredl replied:
I'd be fine with adding a TODO comment for now

shadmansaleh replied:
Ok . I've added a todo . Will look into why it happens later

clason commented:
@shadmansaleh In light if the feature freeze, I'd caution against adding new features here beyond feature parity with (auto)sourcing Lua files from config directories. I'd prefer if we keep this PR at a MVP level and postpone new features to a separate PR (which, if minor enough and non-breaking, can in principle make it into 0.5.1).

shadmansaleh commented:
@clason I'm done . I'll remove the WIP in a while . I basically set it so no one merges the pr while I was waiting on asan report . In case I have some memory errors.

`:source f.lua` had became possible because of a refactor to make this prs implementation simpler . I was looking at how to fully support it . Without supoort for running visually selected lines or lines with `:source` on lua files felt incomplete . So I pushed it here but ofcorce it can go in a separate pr.

clason commented:
Well, the point is that the more complex the feature, the longer we will have to wait after merging before we can release 0.5 to make sure there are no breaking bugs. Before that feature, it looked like purely additive so very unlikely to break existing configs; now I'm no longer so sure.

(This is a general point, not just specific to your PR.)

shadmansaleh commented:
Sorry for increasing complexcity  . But this still shoudn't break anything existing . Unless someone has a file ending with .lua (foo.lua) and has written vimscript in it . The `:source` command now has a check to see if the files name is *.lua . If that check fails the code path is exactly same as before . Since sourceing lua files was an error before . I don't think any config has something like that .

Also I think sourcehuts freebsd CI is broken . Specifically http://www.leonerd.org.uk/ is down it hosts libtermkey source that freebsd CI downloads .

clason commented:
Just making a general point.

And, yes, the outage is known (and unfortunate).

shadmansaleh commented:
@clason also There's probably something wrong with tsan too . It sometimes runs somtimes gets stuck at random locations . I've seen it happen here and in some other prs too . On the feature commit it got stuck . Looking at logs it's stuck on #14249 's commit too https://github.com/neovim/neovim/actions/runs/924340372 .

clason commented:
All builds pull `libtermkey` and `libvterm` from leonerd.org.uk, so they all break currently...

shadmansaleh commented:
@bfredl please take a look at 2a32e444e4e107f7ddb944225832faa3c46a012b and revise your approval . I've pushed a final commit complete support of `:source` on lua files . If you don't find any errors in there this pr is complete :)

shadmansaleh commented:
> All builds pull libtermkey and libvterm from leonerd.org.uk, so they all break currently..

Strange in that case all the CI should break . How did 10 pass :P. The others probably cache downloads right ?

clason commented:
Probably.

bfredl commented:
I agree with the present "cut" of the PR. As I said before my gut feeling is that `:so x.lua` is analogous with `:runtime x.lua`, and the fact that it actually simplifies the implementation of the runtime features confirms that. `:[range]source` is a quite unrelated feature for _end users_ that puns on the same command name (because in this case a pun is better than introducing a new ex command name no one will remember).

shadmansaleh commented:
> :[range]source is a quite unrelated feature for end users that puns on the same command name (because in this case a pun is better than introducing a new ex command name no one will remember).

Yeah . I didn't know it existed before I looked at the code . But after knowing about it I can definitely see uses . With the `:[range]` support `:source` on lua files imitates vimL source completely I think :) .

So I should go to RDY ? Also @bfredl any idea https://github.com/neovim/neovim/pull/14686#discussion_r648781171 ?

shadmansaleh commented:
Ok this PR can be merged if anyone feels like doing it :)

tjdevries commented:
It must be something I have with my environment cause for some reason the files aren't created on my machine when you're doing it in `startup_spec.lua`

```

[ RUN      ] runtime: loads plugin/*.lua from XDG config home: 0.07 ms ERR
test/helpers.lua:777: attempt to index local 'file' (a nil value)

stack traceback:
	test/helpers.lua:777: in function 'write_file'
	test/functional/core/startup_spec.lua:515: in function <test/functional/core/startup_spec.lua:511>
```

I don't love that the new `mkdir_p` uses the current neovim session. I think it should use an os command (like `os_kill` or `do_rmdir`).

But I also don't necessarily want to hold anything up if it's just my environment.

(All the other tests pass for me now, And I don't have any hanging)

shadmansaleh commented:
> I don't love that the new mkdir_p uses the current neovim session. I think it should use an os command (like os_kill or do_rmdir).

Ok I'll write a lua mkdir_p later on :)
